Lib/test/test_complex.py

```diff
@@ -118,6 +122,23 @@ def test_truediv(self):
             self.assertTrue(isnan(z.real))
             self.assertTrue(isnan(z.imag))
 
+        self.assertComplexesAreIdentical(complex('(inf+1j)')/complex('(0+1j)'), complex('(nan-infj)'))
+
+        # test recover of infs if numerator has infinities and denominator is finite
+        self.assertComplexesAreIdentical(complex('(inf-infj)')/(1+0j), complex('(inf-infj)'))
+        self.assertComplexesAreIdentical(complex('(inf+infj)')/(0.0+1j), complex('(inf-infj)'))
+        self.assertComplexesAreIdentical(complex('(nan+infj)')/complex(2**1000, 2**-1000), complex('(inf+infj)'))
```

Comment:
I wonder, how did you came to this test? Its result looks paradoxical at first glance, but makes sense after thinking about it more.

Reply:
That was just a coverage test, that triggers underflow in ``ratio``.  The hard part, as you noted, was to figure out why it's true.  I've just checked the exact answer in the Diofant (probably will work in the SymPy too), e.g. for the real part:
```
In [43]: x = Symbol('x', extended_real=True)
In [44]: y = Symbol('y', positive=True, finite=True)
In [45]: expr = re(expand((x + I*oo)*(y-I/y)))
In [46]: expr.subs({x: Dummy(real=True)})
Out[46]: ∞
In [47]: expr.subs({x: Dummy(positive=True)})
Out[47]: ∞
In [48]: limit(expr, x, -oo)
Out[48]: ∞
```
BTW, Mathematica does more brave things, but hardly it's correct:
```
In[14]:= Hold[(x + I*Infinity)*(2^10-I/2^10)]  (* no assumptions on x ! *)
                                 10    I
Out[14]= Hold[(x + I Infinity) (2   - ---)]
                                       10
                                      2
In[15]:= ReleaseHold[%]

         (1 + 1048576 I) Infinity
Out[15]= ------------------------
           Sqrt[1099511627777]

In[16]:= ReleaseHold[%14/.{x->1-I*Infinity}]

Infinity::indet: Indeterminate expression -I Infinity + I Infinity encountered.

Out[16]= Indeterminate
```

However, it looks I should restore remark about corner cases.  Not all are handled by this code (have no chance, because one component computed as non-nan).  Here is an example:
```pycon
>>> complex('(1+infj)')/complex(2**1000, 2**-1000)
(nan+infj)
```
This is correctly handled by C code both for gcc and clang, but it turns they use a different algorithm for tdiv (not Smith's method, rather something like Priest's algorithm; see full _Cdivd() code before handling nan+nanj case).  Maybe it worth to adopt this, but probably it's out of the scope of this pr.

Reply:
It's quite odd to see NaNs disappearing in an arithmetic operation on operands that include NaNs. However, that seems unavoidable if we want to follow the C standard Annex G semantics, since its rules imply that `complex(inf, nan)` should be regarded as a complex infinity, and then the division of a finite complex number by a complex infinity should be a complex zero (so can't have any NaN components).

Reply:
> It's quite odd to see NaNs disappearing in an arithmetic operation on operands that include NaNs.

I think that could be reasonably interpreted only as shown above, i.e. ``nan`` in components means "anything from the extended real line".

> since its rules imply that complex(inf, nan) should be regarded as a complex infinity

I don't think there is such thing in the C standard.  It has two types of infinities: when one component is finite, nonzero and other is infinite; other type (with few cases) is a directed infinity (phase is finite and absolute value is infinite).  I.e. ``(1+infj)`` and ``(inf+infj)``.  That brings an interesting clash with CAS like Mathematica, that has no notion like the first type if infinities.  E.g. ``ArcSin[Infinity]`` is just ``-I Infinity`` c.f. ``cmath.asin()``.  See https://github.com/mpmath/mpmath/issues/793

But CPython already adopts Annex G semantics in this respect, via ``cmath.isinf()`` and ``cmath.isfinite()``.

@mdickinson, what do you think about changing Smith's algorithm to something more closer to C standard?

Reply:
@skirpichev 

> I don't think there is such thing in the C standard.

Take a look at Annex G. Quoting from §G.3 "Conventions" in C99, paragraph 1:

> A complex or imaginary value with at least one infinite part is regarded as an infinity (even if its other part is a NaN).

The language is slightly changed in the most recent publicly available draft of C23 (n3096.pdf):

> A complex or imaginary value with at least one infinite part is regarded as an infinity (even if its other part is a quiet NaN).